File: final/fecha-2/final.c
// Maximo Santoro. Padron: 105572. DNI: 42822674
// Final - Algoritmos y Programacion II - 5/4/2021

#include <stdlib.h>
#include <stdbool.h>
#include "lista.h"
#include "hash.h"
#include "pila.h"

// // ej 1 - pila
// Se cuenta con un TDA Pila implementado con una estructura enlazada. Implementar en C una primitiva que reciba
// dos pilas, P1 y P2, y cree una nueva pila P3 que tenga todos los elementos de P1, seguidos por todos los elementos de
// P2, en tiempo constante. Tanto P1 como P2 deben quedar destruidas.
// Ejemplo: si tenemos P1 = [8; 7; 5] (el tope es 5) y P2 = [4; 1; 6] (el tope es 6), al finalizar el algoritmo P3 = [8;
// 7; 5; 4; 1; 6] (tope 6). Aclaración: Se puede (y debe) definir cómo es la estructura de la pila para poder cumplir
// con lo pedido, siempre manteniendo que sea una estructura enlazada.

typedef struct nodo {
    void* dato;
    struct nodo* sig;
} nodo_t;

typedef struct pila {
    nodo_t* prim;
    nodo_t* ult;
    size_t largo;
} pila_t;

// defino la pila con una estructura enlazada (con nodos)
// no voy a usar ninguna otra primitiva como "pila_crear()" asi muestro como seria la implementacion interna de cada cosa que se debe hacer. Pero lo optimo seria crear cada primitiva para este tipo de pila enlazada.
pila_t* pila_union(pila_t* p1, pila_t* p2) {
    pila_t* res = malloc(sizeof(pila_t));   // creo la pila
    if (res == NULL) return NULL;   // chequeo que no hayan errores de memoria

    if (p1 == NULL && p2 == NULL) {     // si las dos pilas son NULL, devuelvo una pila enlazada vacia.
        res->largo = 0;
        res->prim = NULL;
        res->ult = NULL;
        // no libero la memoria de las pilas ya que son NULL
        return res; 
    }

    if (p1 == NULL || p1->largo == 0) {   // si p1 es NULL o esta vacia devuelvo una copia de p2. 
        res->largo = p2->largo;
        res->prim = p2->prim;
        res->ult = p2->ult;
        free(p2);
        return res;
    }
    else if (p2 == NULL || p2->largo == 0) {  // si p2 es NULL o esta vacia devuelvo una copia de p1
        res->largo = p1->largo;
        res->prim = p1->prim;
        res->ult = p1->ult;
        free(p1);
        return res;
    }
    // si las 2 pilas son validas y ninguna de las dos esta vacia, entonces procedo a unirlas en la nueva pila "res"
    res->prim = p1->prim;   // el primer nodo de p3 apunta al primer nodo de p1
    p1->ult->sig = p2->prim;    // el ultimo nodo de p1 apunta al primer nodo de p2
    res->ult = p2->ult;     // el ultimo nodo de p3 apunta al ultimo nodo de p2
    res->largo = p1->largo + p2->largo; // sumo los largos
    free(p1);   // destruyo las dos pilas iniciales.
    free(p2);   // lo libero con free porque desconozco de que haria pila_destruir(), ya que no quiero liberar la memoria de los nodos. Solo la de las pilas. Podria usar pila_destruir(p1, NULL); pila_destruir(p2, NULL); pero con el free basta. 
    return res;
}

// Y efectivamente la primitiva anda en tiempo constante (O(1)), ya que lo unico que hago son operaciones de asignacion y comparacion, por ende, todo O(1) e independiente de la cantidad de nodos que tenga cada pila.

// // ej 4

// Implementar una primitiva para el ABB (cuyas claves son números positivos) bool abb_camino_suma(const abb_t*
// abb, int s) que reciba un ABB y un número S y determine si hay un camino dentro del ABB cuyas claves sumen
// exactamente S. Realizar la menor cantidad de comparaciones posibles. Indicar y justificar la complejidad del algoritmo
// implementado.

#include "abb.h"

typedef struct nodo_abb {
    struct nodo_abb *izq;
    struct nodo_abb *der;
    char *clave;
    void *dato;
} nodo_abb_t;

//Definición del struct abb

struct abb {
    nodo_abb_t *raiz;
    size_t cant;
    abb_comparar_clave_t comparar;
    abb_destruir_dato_t destruir;
};

bool abb_camino_suma(const abb_t* abb, int s) {
    if (abb == NULL) return false;
    hash_t* dicc = hash_crear(NULL);
    if (dicc == NULL) return false;
    bool res =  _abb_camino_suma(dicc, abb->raiz, 0, s);
    hash_destruir(dicc);
    return res;
}

bool _abb_camino_suma(hash_t* dicc, nodo_abb_t* nodo, int *suma, int s) {
    if (nodo == NULL) return false;
    if (*suma + (int)(nodo->dato) == s)
        return true;
    else if (*suma + (int)(nodo->dato) > s)
        return false;

    *suma += (int)(nodo->dato);

    if (!hash_guardar(dicc, nodo->clave, nodo->dato)) {
        hash_destruir(dicc);
        return false;
    }

    if (_abb_camino_suma(dicc, nodo->izq, suma, s))
        return true;
    if (_abb_camino_suma(dicc, nodo->der, suma, s))
        return true;
    
    *suma -= (int)(nodo->dato);
    hash_borrar(dicc, nodo->clave);
    return false;
}

// me resulta medio complicado ver la complejidad del algoritmo, pero si no me equivoco, esto en el peor de los casos recorreria una vez todos los elementos del abb. Imaginandome que el numero S exageradamente grande (cosa que nunca pueda llegar a sumar ese numero), voy llamando a cada lado de cada nodo, hasta que ese nodo sea una hoja (o sea NULL). Por lo que terminaria siendo O(n) (aunque estoy podando casos)
// creo que lo puedo resolver con teorema maestro.
// a = 2 -> 2 llamados recursivos
// b = 2 -> proporcion del tamagno original
// c = 0 -> costes extras n^c = 1 -> c = 0
// por ende log 2 (2) = 1, y 1 > 0. 1 > C -> O(n ^ (log B (A))) -> O(n^1) -> O(n)

File: final/fecha-2/final.py
# Maximo Santoro. Padron: 105572. DNI: 42822674
# Final - Algoritmos y Programacion II - 5/4/2021


# # ej 3

# Se tiene un programa que, una vez iniciado, lee números enteros desde entrada estándar (para realizar operaciones que
# son irrelevantes para esta consigna). Una vez que la entrada estándar termina, se desea que el programa informe de los
# K valores más grandes de todos los que se han leído de entrada estándar (supongamos que han sido n números). Es
# importante notar que n podría ser una cantidad que no entre en memoria (e.g.se redirecciona la entrada estándar a
# un archivo de 10 Terabytes). Explicar cómo diseñarías un mecanismo/algoritmo para obtener estos K máximos valores,
# qué estructuras de datos utilizarías para esto, y cuál es la complejidad temporal y espacial de dicho mecanismo/algoritmo.
# Considerar que K (un valor pasado por parámetro al programa) es una cantidad que sí entra en memoria. La complejidad
# (en términos de n y K) debe ser la menor posible, considerando la restricción en memoria. Indicar y justificar dicha
# complejidad.

# La estructura que usaria seria un heap de minimos de tamagno K.
# mi idea es a medida que entran los ints por stdin, ir metiendolos en un Heap de minimos e ir llevando la cuenta de los numeros que entran al Heap. 
# Al principio entrarian los primeros K numeros al heap directo.
# Una vez pasados los primeros K, cada vez que se lea un nuevo numero, me fijo que este sea mas grande que el minimo del heap. 
   # Si es mas grande, saco el minimo del heap, libero su memoria, e ingreso el nuevo numero recien leido al heap. 
   # Si es mas chico o igual al minimo del heap, directamente descarto el numero, y libero su memoria. 

# Una ve hecho este ciclo con los N numeros a leer por entrada standard, tengo mi heap de minimos de tamgno K con los K valores mas GRANDES leidos por stdin. 
# Posteriormente, creo una lista y desencolo del heap K veces appendeando uno por uno a la lista, para que me queden los K numeros ordenados. 
# y finalmente liberaria la memoria de la estructura del heap (no de los datos) y devolveria esta lista como resultado.


# la complejidad de memoria de este programa seria O(K), ya que siempre libero los numeros cuando quedan fuera del heap. Y la lista y las otras cosas que tenga que usar como detalles de implementacion, serian constantes (y despreciables) en terminos de memoria. (la lista usaria la mismos nodos que estan en el heap, asi que su memoria adicional seria cte)

# la complejidad temporal de la lectura e ingreso de numeros quedaria en O(N * log(k)), ya que encolo en el heap (log K), N veces. Y luego una vez leido todo, tendria que desencolar K veces e ingresar esos datos desencolados a una lista. Lo que costaria O(k log K) ya que seria K veces desencolar (O(log K)), y appendear a una lista (O(1))
# Por lo que la complejidad temporal se podria acotar a T(N) = O(N * log(k) + k log (k))



# # ej 2

# Dado un grafo no dirigido y pesado (todos pesos positivos), se desea averiguar el camino mínimo de un vértice v hacia
# todos los demás vértices del grafo. Ahora bien, entre dos pares de vértices v y w podría existir más de un camino
# mínimo posible (por supuesto, todos de mismo costo/peso total). Se desea que ante igualdad de costo, se defina que el
# camino a tomar sea el más corto en cantidad de aristas. Explicar (en texto y/o implementando) cómo modificarías
# el algoritmo de Dijkstra para cumplir lo pedido. La complejidad del algoritmo debería quedar tal cual el algoritmo
# original. Indicar y justificar cuál es dicha complejidad, además de por qué no se modifica con la solución planteada.

# Ejemplo: Para el siguiente grafo, se puede llegar desde A hacia D con 3 caminos posibles (detallados debajo). Todos de
# costo 10. Se debe asegurar que el camino elegido sea el primero (el más corto). Lo mismo para los demás vértices destino.

# RTA

# la complejidad de dijkstra es O(E * log V), siendo E la cantidad de aristas, y V la cantidad de vertices

# lo que yo agregaria al algoritmo seria un diccionario de orden, y cuando se encuentra un camino de igual peso, me fijo que el orden sea menor. La complejidad del algoritmo no deberia variar.
# No deberia variar ya que seria unicamente agregar operaciones constantes, o sea, comparaciones (un if), y la creacion y agregar datos al diccionario de orden




# # ej 5
# En clase se ha demostrado que ningún algoritmo de ordenamiento comparativo puede ser mejor que O(n log n), siendo n
# la cantidad de elementos del arreglo a ordenar. Al mismo tiempo, hemos visto algoritmos para obtener un ordenamiento
# topológico de un conjunto de datos que, modelando el problema con grafos (dirigidos), termina resolviéndose en tiempo
# lineal a la cantidad de vértices y aristas del mismo. Explicar por qué en problemas de ordenamientos (e.g.arreglo de
# números de los que no se tiene información) no modelamos siempre a estos con grafos y luego resolvemos utilizando
# alguno de los algoritmos antes mencionados.
# Recomendación: pensar bien este ejercicio, hacerse dibujos, etc. . . antes de ponerse a escribir la respuesta. No responder
# “no se puede resolver de esta forma”, porque sí se puede. Por el contrario, para plantear la respuesta conviene plantearse
# cómo se haría para resolver un problema de ordenamiento usando una resolución de orden topológico. Esta consigna no
# va sobre la posibilidad, sino por qué no es la mejor opción para hacerlo, a pesar de parecer una idea tentadora.

# RTA:
# yo creo que el gran problema a la hora de encarar un problema de ordenamiento de esta forma, seria la complejidad temporal de modelar el grafo. teniendo en cuenta que deberiamos recorrer todo el arreglo de numeros e ir armando el grafo.
# 
# #primero tendria que recorrer todo el arreglo y agregar cada elemento al grafo -> O(n) 
# Y luego tendria que armar las aristas.
# Esto lo haria buscando el minimo vertice en todo el grafo, y una vez obtenido V, iria yendo vertice por vertice indicando cuando hay una arista desde ese vertice hacia otro. (lo cual yo lo haria buscando el proximo numero inmediato ordenado en el grafo, o sea si tengo [1, 3, 4, 8, 9, 2]. arranco por el 1, recorreria todos los vertices hasta encontrar el minimo numero mayor o igual a 1 para conectarlo.)
# Mantendria un diccionario de visitados, y continuaria haciendo lo mismo con el numero al cual conecte el vertice anterior. En el ejemplo dado, conectaria 1->2 ... luego con el 2, busco el proximo minimo mayor o igual a 2. Y asi iria conectando. 
# asi sucesivamente hasta completar todas las conecciones del grafo, y me terminaria dando algo similar a O(N^2) ya que recorro N N veces. y luego, quedaria la parte del orden topologico, o sea, O(n + e). Por lo que creo que no convendria usar este metodo. Ya que me quedaria O N^2. El problema reside en la modelacion del grafo. Supongo que hay una forma mejor de modelarlo de lo que presente, pero estoy seguro que sin informacion previa del arreglo y sus elementos es imposible lograr hacer un modelado que termine siendo mejor que O(n log n). Por lo que seguramente se pueda hacer, pero no valdria la pena nunca (siempre y cuando NO tengamos informacion adicional del arreglo a ordenar o sus elementos. )